renderdoc/android/android.cpp

```diff
@@ -487,31 +528,34 @@ RDResult InstallRenderDocServer(const rdcstr &deviceID)
 
     RDCLOG("Installed package '%s', checking for success...", apk.c_str());
 
-    AndroidVersionCheckResult check = CheckAndroidServerVersion(deviceID, abi);
+    AndroidVersionCheckResult versionCheck = CheckAndroidServerVersion(deviceID, abi);
+    AndroidInstallPermissionCheckResult permissionsCheck =
+        CheckAndroidServerInstallPermissions(deviceID, GetRenderDocPackageForABI(abi), apiVersion);
 
-    if(check != AndroidVersionCheckResult::Correct)
+    if((versionCheck != AndroidVersionCheckResult::Correct) ||
+       (permissionsCheck != AndroidInstallPermissionCheckResult::Correct))
     {
       RDCLOG("Failed to install APK. stdout: %s, stderr: %s",
              adbInstall.strStdout.trimmed().c_str(), adbInstall.strStderror.trimmed().c_str());
       RDCLOG("Retrying...");
-      adbExecCommand(deviceID, "install -r \"" + apk + "\"");
+      InstallAPK(deviceID, apk, apiVersion);
```

Comment:
Query in terms of a change and if it makes any difference

The old code would use `install -r` for the retry of the install but the new code will use a different install option ie. `install -r -g` or `install -r -g --force-queryable`

I am wondering if this change has any impact, on the face of it perhaps the old code was incorrect in the way it tried to do the second install

Reply:
There is no drawback to adding `-g (--force-queryable)` on reinstall, even if all permissions were granted first time round - so the code was not incorrect per se, it just assumes `-g (--force-queryable)` worked the first time. There should be no change in behavior there.

With regards to this ticket, on affected devices the initial `-g` flag would fail to grant all requested permissions on first install, so a reinstall in the case of `permissionsCheck` failing does actually need both flags. For simplicity, I have kept a combined retry mechanism (even if the flags are redundant in some cases).

Reply:
I don't fully agree with it having no drawbacks. **Every** adb command or flag is a potential failure point because it might cause things to break on some device, given there's no guarantee that any adb commands behave the same or correctly across android.

The second install not passing `-g` was deliberate, to try to handle the case of devices that failed on the first adb command because of the presence of unhandled flags. It would fall back to a potentially safer install (and then users need to grant the permissions manually - which they need to do for some permissions anyway). Since this PR is a workaround is for broken devices it should not change the behaviour more than necessary, even for other workarounds for other broken devices.

Reply:
Bumping this an unresolving the conversation - I thought it would be reopened when I made a review comment but I think it may have been missed as it stayed collapsed.

Reply:
I have updated this section to keep the `-r` only fallback as before in the case of a version check failure. Permission check now happens in isolation to avoid side effects on other devices. I am happy to iterate on this as needed.